rust/worker/src/distance/types.rs

```diff
@@ -43,6 +92,32 @@ impl DistanceFunction {
                 1.0_f32 - sum
             }
             DistanceFunction::InnerProduct => {
+                #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+                {
+                    if std::arch::is_aarch64_feature_detected!("neon") {
+                        return unsafe { crate::distance::distance_neon::inner_product(a, b) };
+                    }
+                }
+                #[cfg(all(
+                    any(target_arch = "x86_64", target_arch = "x86"),
+                    target_feature = "sse"
+                ))]
+                {
+                    if std::arch::is_x86_feature_detected!("sse") {
+                        return unsafe { crate::distance::distance_sse::inner_product(a, b) };
+                    }
+                }
+                #[cfg(all(
+                    target_arch = "x86_64",
+                    all(target_feature = "avx", target_feature = "fma")
+                ))]
+                {
+                    if std::arch::is_x86_feature_detected!("avx")
```

Comment:
should we add tests to validate that the simd impls (whatever can run on the target machine) match the base impl?

Reply:
I think there is a test already in `types.rs` - `test_distance_function_l2sqr()` which I was relying on. If that regresses then there is some bug in the SIMD impls. That test validates inner product and l2 norm. Since, cosine is the same as inner product for us, I am guessing that is not needed

Reply:
Ok great good point